// venom-bot/dist/controllers/browser.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getWhatsappPage = exports.initBrowser = exports.folderSession = exports.puppeteerMutationListener = exports.statusLog = exports.initWhatsapp = void 0;
const ChromeLauncher = __importStar(require("chrome-launcher"));
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const puppeteer_extra_1 = __importDefault(require("puppeteer-extra"));
const puppeteer_config_1 = require("../config/puppeteer.config");
const StealthPlugin = require("puppeteer-extra-plugin-stealth");
const auth_1 = require("./auth");
const WAuserAgente_1 = require("../config/WAuserAgente");
const sleep_1 = require("../utils/sleep");
const os = __importStar(require("os"));
const rimraf = __importStar(require("rimraf"));
async function initWhatsapp(session, options, browser, token) {
    const waPage = await getWhatsappPage(browser);
    if (waPage != null) {
        try {
            await waPage.setUserAgent(WAuserAgente_1.useragentOverride);
            //await waPage.setCacheEnabled(true);
            if (typeof options.userPass === 'string' &&
                options.userPass.length &&
                typeof options.userProxy === 'string' &&
                options.userProxy.length &&
                Array.isArray(options.addProxy) &&
                options.addProxy.length) {
                await waPage.authenticate({
                    username: options.userProxy,
                    password: options.userPass
                });
            }
            await waPage.goto(puppeteer_config_1.puppeteerConfig.whatsappUrl, {
                waitUntil: 'domcontentloaded'
            });
            waPage.on('pageerror', ({ message }) => {
                const erroLog = message.includes('RegisterEffect is not a function');
                if (erroLog) {
                    waPage.evaluate(() => {
                        localStorage.clear();
                        window.location.reload();
                    });
                }
            });
            await browser.userAgent();
            // Auth with token
            await (0, auth_1.auth_InjectToken)(waPage, session, options, token);
            return waPage;
        }
        catch (e) {
            waPage.close().catch(() => { });
            browser.close().catch(() => { });
            return false;
        }
    }
}
exports.initWhatsapp = initWhatsapp;
async function statusLog(page, spinnies, session, callback) {
    while (true) {
        if (page.isClosed()) {
            try {
                spinnies.fail(`whatzapp-intro-${session}`, {
                    text: 'Erro intro'
                });
            }
            catch (_a) { }
            break;
        }
        const infoLog = await page
            .evaluate(() => {
            const target = document.getElementsByClassName('_2dfCc');
            if (target && target.length) {
                if (target[0]['innerText'] !== 'WhatsApp' &&
                    target[0]['innerText'] !== window['statusInicial']) {
                    window['statusInicial'] = target[0]['innerText'];
                    return window['statusInicial'];
                }
            }
        })
            .catch(() => undefined);
        if (infoLog) {
            callback(infoLog);
        }
        await (0, sleep_1.sleep)(200);
    }
}
exports.statusLog = statusLog;
async function puppeteerMutationListener(oldValue, newValue) {
    console.log(`${oldValue} -> ${newValue}`);
}
exports.puppeteerMutationListener = puppeteerMutationListener;
function folderSession(options, session, returnFolder = true) {
    const folderSession = path.join(path.resolve(process.cwd(), options.mkdirFolderToken, options.folderNameToken, options.mkdirFolderTokenMultidevice, session));
    if (!fs.existsSync(folderSession)) {
        fs.mkdirSync(folderSession, {
            recursive: true
        });
    }
    const folderMulidevice = path.join(path.resolve(process.cwd(), options.mkdirFolderToken, options.folderNameToken));
    if (!fs.existsSync(folderMulidevice)) {
        fs.mkdirSync(folderMulidevice, {
            recursive: true
        });
    }
    fs.chmodSync(folderMulidevice, '777');
    fs.chmodSync(folderSession, '777');
    options.puppeteerOptions = {
        userDataDir: folderSession,
        ignoreHTTPSErrors: true
    };
    puppeteer_config_1.puppeteerConfig.chromiumArgs.push(`--user-data-dir=${folderSession}`);
    if (returnFolder) {
        return folderSession;
    }
}
exports.folderSession = folderSession;
/**
 * Initializes browser, will try to use chrome as default
 * @param session
 */
async function initBrowser(spinnies, session, options, extras = {}) {
    const chromePath = getChrome();
    if (chromePath && options.useChrome) {
        extras = { ...extras, executablePath: chromePath };
    }
    else {
        if (options.BrowserFetcher) {
            const browserFetcher = puppeteer_extra_1.default.createBrowserFetcher(undefined);
            spinnies.add(`extract-file`, {
                text: `Await download Chromium`
            });
            let init = true;
            await browserFetcher
                .download(options.chromiumVersion, (downloadedByte, totalBytes) => {
                if (init) {
                    spinnies.add(`chromium-${session}`, {
                        text: 'Check chromium....'
                    });
                    spinnies.add(`chromium-total-${session}`, {
                        text: 'Checking the total bytes to download!'
                    });
                    init = false;
                }
                if (downloadedByte) {
                    spinnies.update(`chromium-${session}`, {
                        text: `Total byte value....${downloadedByte}`
                    });
                }
                if (totalBytes) {
                    spinnies.update(`chromium-total-${session}`, {
                        text: `Total Bytes ${totalBytes}`
                    });
                }
                if (downloadedByte === totalBytes) {
                    spinnies.succeed(`chromium-${session}`, {
                        text: `Chromium Finished result`
                    });
                    spinnies.succeed(`chromium-total-${session}`, {
                        text: `Chromium completed result`
                    });
                    spinnies.update(`extract-file`, {
                        text: `Extract files... await...`
                    });
                }
            })
                .then((revisionInfo) => {
                spinnies.succeed(`extract-file`, {
                    text: `Extract files concluded`
                });
                extras = {
                    ...extras,
                    executablePath: revisionInfo.executablePath
                };
                puppeteer_config_1.puppeteerConfig.chromiumArgs.push(`--single-process`);
            })
                .catch((e) => {
                console.log('error chromium: ', e);
                extras = {};
            });
        }
        else {
            console.log('Chrome not found, using chromium');
            extras = {};
        }
    }
    if (options.multidevice === true) {
        folderSession(options, session);
    }
    // Use stealth plugin to avoid being detected as a bot
    puppeteer_extra_1.default.use(StealthPlugin());
    if (Array.isArray(options.addProxy) && options.addProxy.length) {
        const proxy = options.addProxy[Math.floor(Math.random() * options.addProxy.length)];
        options.browserArgs
            ? Object.assign(options.browserArgs, [`--proxy-server=${proxy}`])
            : Object.assign(puppeteer_config_1.puppeteerConfig.chromiumArgs, [
                `--proxy-server=${proxy}`
            ]);
        // console.log(puppeteerConfig.chromiumArgs);
    }
    if (Array.isArray(options === null || options === void 0 ? void 0 : options.addBrowserArgs) &&
        (options === null || options === void 0 ? void 0 : options.addBrowserArgs.length) > 0) {
        for (let index = 0; index < (options === null || options === void 0 ? void 0 : options.addBrowserArgs.length); index++) {
            const element = options === null || options === void 0 ? void 0 : options.addBrowserArgs[index];
            if (!puppeteer_config_1.puppeteerConfig.chromiumArgs.includes(element)) {
                puppeteer_config_1.puppeteerConfig.chromiumArgs.push(element);
            }
        }
    }
    let browser = null;
    if (options.browserWS && options.browserWS != '') {
        await puppeteer_extra_1.default
            .connect({
            browserWSEndpoint: options.browserWS
        })
            .then((e) => {
            browser = e;
        })
            .catch(() => {
            browser = 'connect';
        });
    }
    else {
        await puppeteer_extra_1.default
            .launch({
            headless: options.headless,
            devtools: options.devtools,
            args: Array.isArray(options.browserArgs) && options.browserArgs.length
                ? options.browserArgs
                : [...puppeteer_config_1.puppeteerConfig.chromiumArgs],
            ...options.puppeteerOptions,
            ...extras
        })
            .then((e) => {
            browser = e;
        })
            .catch((e) => {
            console.error('Error launch: ', e);
            browser = 'launch';
        });
    }
    try {
        const arg = browser
            .process()
            .spawnargs.find((s) => s.startsWith('--user-data-dir='));
        if (arg) {
            const tmpUserDataDir = arg.split('=')[1];
            // Only if path is in TMP directory
            if (path.relative(os.tmpdir(), tmpUserDataDir).startsWith('puppeteer')) {
                process.on('exit', () => {
                    // Remove only on exit signal
                    try {
                        rimraf.sync(tmpUserDataDir);
                    }
                    catch (error) { }
                });
            }
        }
    }
    catch (error) { }
    return browser;
}
exports.initBrowser = initBrowser;
async function getWhatsappPage(browser) {
    const pages = await browser.pages().catch();
    if (pages.length) {
        return pages[0];
    }
    return await browser.newPage().catch();
}
exports.getWhatsappPage = getWhatsappPage;
/**
 * Retrieves chrome instance path
 */
function getChrome() {
    try {
        const chromeInstalations = ChromeLauncher.Launcher.getInstallations();
        return chromeInstalations[0];
    }
    catch (error) {
        return undefined;
    }
}
