// venom-bot/dist/controllers/initializer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.create = exports.InterfaceStateChange = void 0;
const fs_1 = require("fs");
const whatsapp_1 = require("../api/whatsapp");
const create_config_1 = require("../config/create-config");
const auth_1 = require("./auth");
const browser_1 = require("./browser");
const welcome_1 = require("./welcome");
const spinnies_1 = require("../utils/spinnies");
const enum_1 = require("../api/model/enum");
const helpers_1 = require("../api/helpers");
const check_up_to_date_1 = require("./check-up-to-date");
const QRCode = require('qrcode');
var InterfaceStateChange;
(function (InterfaceStateChange) {
    /**
     * Client interface is loading page from qrcode
     */
    InterfaceStateChange["qrcodeOpening"] = "qrcodeOpening";
    /**
     * Client interface is loading qrcode
     */
    InterfaceStateChange["qrcodeLoading"] = "qrcodeLoading";
    /**
     * QR code ready to be read!
     */
    InterfaceStateChange["qrcodeNormal"] = "qrcodeNormal";
    /**
     * Client interface is loading page from syncing
     */
    InterfaceStateChange["syncingOpening"] = "syncingOpening";
    /**
     * Client interface is loading syncing
     */
    InterfaceStateChange["syncingLoading"] = "syncingLoading";
    /**
     * Syncing ready to be read!
     */
    InterfaceStateChange["syncingNormal"] = "syncingNormal";
    /**
     * The customer is in the chat
     */
    InterfaceStateChange["chatsAvailable"] = "chatsAvailable";
})(InterfaceStateChange = exports.InterfaceStateChange || (exports.InterfaceStateChange = {}));
/**
 * @deprecated depreciated due to Multidevices {@link browserSessionToken}
 */
async function create(sessionOrOption, catchQR, statusFind, options, browserSessionToken, browserInstance, reconnectQrcode, interfaceChange) {
    let session = 'session';
    return new Promise(async (resolve, reject) => {
        if (typeof sessionOrOption === 'string' &&
            sessionOrOption.replace(/\s/g, '').length) {
            session = sessionOrOption.replace(/\s/g, '');
        }
        else if (typeof sessionOrOption === 'object') {
            session = sessionOrOption.session || session;
            catchQR = sessionOrOption.catchQR || catchQR;
            statusFind = sessionOrOption.statusFind || statusFind;
            browserSessionToken =
                sessionOrOption.browserSessionToken || browserSessionToken;
            browserInstance = sessionOrOption.browserInstance || browserInstance;
            options = sessionOrOption;
        }
        await (0, check_up_to_date_1.checkUpdates)();
        let browserToken;
        const spinnies = (0, spinnies_1.getSpinnies)({
            disableSpins: options ? options.disableSpins : false
        });
        const mergedOptions = { ...create_config_1.defaultOptions, ...options };
        if (!mergedOptions.disableWelcome) {
            (0, welcome_1.welcomeScreen)();
        }
        statusFind && statusFind('initBrowser', session);
        // Initialize whatsapp
        if (mergedOptions.browserWS) {
            spinnies.add(`browser-${session}`, {
                text: `Waiting... checking the wss server...`
            });
        }
        else {
            spinnies.add(`browser-${session}`, {
                text: 'Waiting... checking the browser...'
            });
        }
        const browser = await (0, browser_1.initBrowser)(spinnies, session, mergedOptions);
        // Erro of connect wss
        if (typeof browser === 'string' && browser === 'connect') {
            spinnies.fail(`browser-${session}`, {
                text: `Error when try to connect ${mergedOptions.browserWS}`
            });
            statusFind && statusFind('serverWssNotConnected', session);
            return reject(`Error when try to connect ${mergedOptions.browserWS}`);
        }
        // Erro open browser
        if (typeof browser === 'string' && browser === 'launch') {
            spinnies.fail(`browser-${session}`, {
                text: `Error no open browser....`
            });
            statusFind && statusFind('noOpenBrowser', session);
            return reject(`Error no open browser....`);
        }
        if (mergedOptions.browserWS) {
            statusFind && statusFind('connectBrowserWs', session);
            spinnies.succeed(`browser-${session}`, {
                text: `Has been properly connected to the wss server`
            });
        }
        else {
            statusFind && statusFind('openBrowser', session);
            spinnies.succeed(`browser-${session}`, {
                text: `Browser successfully opened`
            });
        }
        if (!mergedOptions.browserWS) {
            spinnies.add(`browser-${session}`, {
                text: 'checking headless...'
            });
            if (mergedOptions.headless) {
                spinnies.succeed(`browser-${session}`, {
                    text: 'headless option is active, browser hidden'
                });
            }
            else {
                spinnies.succeed(`browser-${session}`, {
                    text: 'headless option is disabled, browser visible'
                });
            }
        }
        if (typeof browser === 'object') {
            if (!mergedOptions.browserWS && browser['_process']) {
                browser['_process'].once('close', () => {
                    browser['isClose'] = true;
                });
            }
            (0, helpers_1.checkingCloses)(browser, mergedOptions, (result) => {
                statusFind && statusFind(result, session);
            }).catch(() => {
                spinnies.fail(`whatzapp-${session}-close`, {
                    text: 'Closed Browser'
                });
                return reject('The client has been closed');
            });
            if ((0, auth_1.SessionTokenCkeck)(browserSessionToken)) {
                browserToken = browserSessionToken;
            }
            spinnies.add(`whatzapp-${session}`, {
                text: 'Checking page to whatzapp...'
            });
            statusFind && statusFind('initWhatsapp', session);
            // Initialize whatsapp
            const page = await (0, browser_1.initWhatsapp)(session, mergedOptions, browser, browserToken);
            if (page === false) {
                spinnies.fail(`whatzapp-${session}`, {
                    text: 'Error accessing the page: "https://web.whatsapp.com"'
                });
                statusFind && statusFind('erroPageWhatsapp', session);
                return reject('Error when trying to access the page: "https://web.whatsapp.com"');
            }
            statusFind && statusFind('successPageWhatsapp', session);
            spinnies.succeed(`whatzapp-${session}`, {
                text: 'Page successfully accessed'
            });
            try {
                spinnies.add(`whatzapp-intro-${session}`, {
                    text: 'waiting for introduction'
                });
            }
            catch (_a) { }
            (0, browser_1.statusLog)(page, spinnies, session, (event) => {
                try {
                    spinnies.add(`whatzapp-intro-${session}`, {
                        text: event
                    });
                }
                catch (_a) { }
                statusFind && statusFind('introductionHistory', session, event);
            });
            // if (mergedOptions.forceConnect) {
            //   loadForceConnect(
            //     page,
            //     async (event) => {
            //       if (typeof event === 'boolean' && event === true) {
            //         statusFind && statusFind('executedLoadForce', session);
            //       } else {
            //         statusFind && statusFind('infoLoadForce', session, event);
            //       }
            //     },
            //     mergedOptions.attemptsForceConnectLoad,
            //     mergedOptions.forceConnectTime
            //   ).catch();
            // }
            const client = new whatsapp_1.Whatsapp(browser, page, session, mergedOptions);
            //client.initialize();
            if (browserInstance) {
                browserInstance(browser, page, client);
            }
            // checkStore(page, client);
            client.onInterfaceChange(async (interFace) => {
                try {
                    if (interFace.mode === enum_1.InterfaceMode.MAIN) {
                        interfaceChange && interfaceChange('chatsAvailable', session);
                        //checkDisconnect(page, client);
                        spinnies.add(`whatzapp-mode-main-${session}`, {
                            text: 'opening main page...'
                        });
                        spinnies.succeed(`whatzapp-mode-main-${session}`, {
                            text: 'Successfully main page!'
                        });
                        spinnies.succeed(`whatzapp-mode-syncing-${session}`, {
                            text: 'Successfully sync!'
                        });
                        await client.initService();
                        await client.addChatWapi();
                    }
                    if (interFace.mode === enum_1.InterfaceMode.SYNCING) {
                        if (interFace.info === enum_1.InterfaceState.OPENING) {
                            interfaceChange && interfaceChange('syncingOpening', session);
                            spinnies.add(`whatzapp-mode-syncing-${session}`, {
                                text: 'opening sync page...'
                            });
                        }
                        if (interFace.info === enum_1.InterfaceState.PAIRING) {
                            interfaceChange && interfaceChange('syncingLoading', session);
                            spinnies.add(`whatzapp-mode-syncing-${session}`, {
                                text: 'Loading sync...'
                            });
                        }
                        if (interFace.info === enum_1.InterfaceState.NORMAL) {
                            interfaceChange && interfaceChange('syncingNormal', session);
                            spinnies.succeed(`whatzapp-mode-syncing-${session}`, {
                                text: 'Successfully sync!'
                            });
                        }
                    }
                    if (interFace.mode === enum_1.InterfaceMode.QR) {
                        const status = await page.evaluate(() => { var _a, _b, _c; return (_c = (_b = (_a = window === null || window === void 0 ? void 0 : window.Store) === null || _a === void 0 ? void 0 : _a.State) === null || _b === void 0 ? void 0 : _b.Socket) === null || _c === void 0 ? void 0 : _c.stream; });
                        if (status === enum_1.SocketStream.DISCONNECTED) {
                            spinnies.add(`whatzapp-disconnected-${session}`, {
                                text: 'checking...'
                            });
                            spinnies.fail(`whatzapp-disconnected-${session}`, {
                                text: 'Was disconnected!'
                            });
                            statusFind && statusFind('desconnected', session);
                        }
                        if (interFace.info === enum_1.InterfaceState.OPENING) {
                            interfaceChange && interfaceChange('qrcodeOpening', session);
                            spinnies.add(`whatzapp-mode-qr-${session}`, {
                                text: 'Opening QR Code page...'
                            });
                        }
                        if (interFace.info === enum_1.InterfaceState.PAIRING) {
                            interfaceChange && interfaceChange('qrcodeLoading', session);
                            spinnies.add(`whatzapp-mode-qr-${session}`, {
                                text: 'Loading QR Code...'
                            });
                        }
                        if (interFace.info === enum_1.InterfaceState.NORMAL) {
                            interfaceChange && interfaceChange('qrcodeNormal', session);
                            spinnies.succeed(`whatzapp-mode-qr-${session}`, {
                                text: 'Successfully loaded QR Code!'
                            });
                        }
                    }
                }
                catch (_a) { }
            });
            client
                .onStreamChange(async (stateStream) => {
                if (stateStream === enum_1.SocketStream.CONNECTED) {
                    try {
                        spinnies.succeed(`whatzapp-intro-${session}`, {
                            text: 'Successfully connected!'
                        });
                    }
                    catch (_a) { }
                }
                // if (mergedOptions.createPathFileToken) {
                //   if (stateStream === SocketStream.CONNECTED) {
                //     saveToken(page, session, mergedOptions);
                //   }
                // }
                if (stateStream === enum_1.SocketStream.DISCONNECTED) {
                    const mode = await page
                        .evaluate(() => { var _a, _b; return (_b = (_a = window === null || window === void 0 ? void 0 : window.Store) === null || _a === void 0 ? void 0 : _a.Stream) === null || _b === void 0 ? void 0 : _b.mode; })
                        .catch(() => { });
                    if (mode === enum_1.InterfaceMode.QR
                    // && checkFileJson(mergedOptions, session)
                    ) {
                        if (statusFind) {
                            spinnies.add(`whatzapp-qr-${session}`, {
                                text: 'check....'
                            });
                            statusFind('desconnectedMobile', session);
                            spinnies.fail(`whatzapp-qr-${session}`, {
                                text: 'Disconnected by cell phone!'
                            });
                        }
                        //deleteFiles(mergedOptions, session, spinnies);
                    }
                }
            })
                .catch();
            client
                .onStateChange(async (state) => {
                if (state === enum_1.SocketState.PAIRING) {
                    const device = await page
                        .evaluate(() => {
                        var _a, _b, _c, _d;
                        if (document.querySelector('[tabindex="-1"]') &&
                            ((_b = (_a = window === null || window === void 0 ? void 0 : window.Store) === null || _a === void 0 ? void 0 : _a.Stream) === null || _b === void 0 ? void 0 : _b.mode) === enum_1.InterfaceMode.SYNCING &&
                            ((_d = (_c = window === null || window === void 0 ? void 0 : window.Store) === null || _c === void 0 ? void 0 : _c.Stream) === null || _d === void 0 ? void 0 : _d.obscurity) === 'SHOW') {
                            return true;
                        }
                        return false;
                    })
                        .catch(() => undefined);
                    if (device === true) {
                        const ckeckVersion = await (0, auth_1.isClassic)(page);
                        if (ckeckVersion === false) {
                            await page.evaluate(async () => {
                                var _a, _b;
                                await ((_b = (_a = window === null || window === void 0 ? void 0 : window.Store) === null || _a === void 0 ? void 0 : _a.Login) === null || _b === void 0 ? void 0 : _b.triggerCriticalSyncLogout());
                            });
                        }
                        if (statusFind) {
                            statusFind('deviceNotConnected', session);
                        }
                    }
                }
            })
                .catch();
            page.on('dialog', async (dialog) => {
                await dialog.accept();
            });
            if (mergedOptions.waitForLogin) {
                const isLogged = await client
                    .waitForLogin(catchQR, statusFind)
                    .catch(() => undefined);
                statusFind && statusFind('waitForLogin', session);
                if (!isLogged) {
                    return reject('Not Logged');
                }
                let waitLoginPromise = null;
                client
                    .onStateChange(async (state) => {
                    if (state === enum_1.SocketState.UNPAIRED ||
                        state === enum_1.SocketState.UNPAIRED_IDLE) {
                        if (!waitLoginPromise) {
                            waitLoginPromise = client
                                .waitForLogin(catchQR, statusFind)
                                .then(() => {
                                if (reconnectQrcode) {
                                    reconnectQrcode(client);
                                }
                            })
                                .catch(() => { })
                                .finally(() => {
                                waitLoginPromise = null;
                            });
                        }
                        await waitLoginPromise;
                    }
                })
                    .catch();
            }
            if (mergedOptions.debug) {
                const debugURL = `http://localhost:${(0, fs_1.readFileSync)(`./${session}/DevToolsActivePort`).slice(0, -54)}`;
                console.log(`\nDebug: \x1b[34m${debugURL}\x1b[0m`);
            }
            statusFind && statusFind('waitChat', session);
            await page
                .waitForSelector('#app .two', { visible: true })
                .catch(() => { });
            await page
                .waitForFunction(() => {
                if (mergedOptions.debug) {
                    console.log(`\nDebug: Loading wp app....`);
                }
                const StoreKey = Object.keys(window).find((k) => !!Object.getOwnPropertyDescriptor(window[k], 'WidFactory') &&
                    !!Object.getOwnPropertyDescriptor(window[k].WidFactory, 'createWid'));
                if (StoreKey) {
                    window.Store = window[StoreKey];
                    return true;
                }
                return false;
            }, {
                timeout: 0,
                polling: 100
            })
                .catch(() => { });
            try {
                spinnies.succeed(`whatzapp-intro-${session}`, {
                    text: 'Successfully connected!'
                });
            }
            catch (_b) { }
            await client.initService();
            await client.addChatWapi();
            // resetStore(page);
            //await sleep(2000);
            statusFind && statusFind('successChat', session);
            // const atualizar = await page.waitForSelector('span[data-testid="chat-butterbar"]').catch(() => { });
            // if (atualizar) {
            //   await page.evaluate(() => {
            //     window.updater.restart()
            //   });
            //   await page
            //     .waitForSelector('#app .two', { visible: true })
            //     .catch(() => { });
            // }
            return resolve(client);
        }
    });
}
exports.create = create;
